features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64LRUCache.kt

```diff
@@ -0,0 +1,135 @@
+
+import android.content.ComponentCallbacks2
+import android.content.res.Configuration
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.DrawableContainer
+import android.graphics.drawable.LayerDrawable
+import android.util.LruCache
+import android.view.View
+import androidx.annotation.VisibleForTesting
+
+@Suppress("MagicNumber")
+internal object Base64LRUCache : Cache, ComponentCallbacks2 {
+    private const val MAX_CACHE_MEMORY_SIZE_BYTES = 4 * 1024 * 1024 // 4MB
+    private const val ON_LOW_MEMORY_SIZE_BYTES = MAX_CACHE_MEMORY_SIZE_BYTES / 2 // 50% size
+    private const val ON_MODERATE_MEMORY_SIZE_BYTES = (MAX_CACHE_MEMORY_SIZE_BYTES / 4) * 3 // 75% size
+    private var componentCallbacksRegistered = false
+
+    private var cache: LruCache<String, ByteArray> = object :
+        LruCache<String, ByteArray>(MAX_CACHE_MEMORY_SIZE_BYTES) {
+        override fun sizeOf(key: String?, value: ByteArray): Int {
+            return value.size
+        }
+    }
+
+    override fun onTrimMemory(level: Int) {
+        when (level) {
+            ComponentCallbacks2.TRIM_MEMORY_BACKGROUND -> {
+                cache.evictAll()
+            }
+
+            ComponentCallbacks2.TRIM_MEMORY_COMPLETE -> {
+                cache.evictAll()
+            }
+
+            ComponentCallbacks2.TRIM_MEMORY_MODERATE -> {
+                cache.trimToSize(ON_MODERATE_MEMORY_SIZE_BYTES)
+            }
+
+            ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL -> {
+                cache.evictAll()
+            }
+
+            ComponentCallbacks2.TRIM_MEMORY_RUNNING_LOW -> {
+                cache.trimToSize(ON_LOW_MEMORY_SIZE_BYTES)
+            }
+
+            ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE -> {
+                cache.trimToSize(ON_MODERATE_MEMORY_SIZE_BYTES)
+            }
+
+            ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN -> {
+                cache.evictAll()
+            }
+
+            else -> {
+                cache.evictAll()
+            }
+        }
+    }
+
+    override fun registerComponentCallbacksIfNecessary(view: View) {
+        if (!componentCallbacksRegistered) {
+            view.context.applicationContext.registerComponentCallbacks(this)
+            componentCallbacksRegistered = true
+        }
+    }
+
+    override fun onConfigurationChanged(newConfig: Configuration) {}
+
+    override fun onLowMemory() {
+        cache.evictAll()
+    }
+
+    @VisibleForTesting
+    internal fun setBackingCache(cache: LruCache<String, ByteArray>) {
+        this.cache = cache
+    }
+
+    @Synchronized
+    override fun put(drawable: Drawable, base64String: String) {
+        val key = generateKey(drawable)
+        val value = base64String.toByteArray()
+        cache.put(key, value)
+    }
+
+    @Synchronized
+    override fun get(drawable: Drawable): String? =
+        cache.get(generateKey(drawable))?.let {
+            String(it)
```

Comment:
seems like unnecessary allocation here - we don't need a string down the line, because anyway we store all the data in the byte array, so I guess we create it here for nothing. It is better to return `ByteArray` from here directly imo.

Reply:
I'm returning `String` here because we're using that eventually in the wireframe property, however I can change to`ByteArray` so that the cache won't perform the conversion